### PR TITLE
fix ksmc_suspendEnvironment() deaklock

### PR DESCRIPTION
**Pull request by JunyiXie**
after suspend all threads
vm_deallocate may cause deadlock

### CODE REVIEW

**Comment by ZhiyuWong** [comment]
That's indeed a problem, but use `threads` after deallocated maybe dangerous.
In my point of view, use a global variable to reference suspended `threads` and deallocate it when `ksmc_resumeEnvironment()` called should be better.

**Comment by JunyiXie** [comment]
Use an array of size 1000 to store threads

**Comment by JunyiXie** [comment]
Add a WeChat? @ZhiyuWong

File: Source/KSCrash/Recording/Tools/KSMachineContext.c [hunk 1]
```diff
@@ -182,9 +184,21 @@ void ksmc_suspendEnvironment()
         return;
     }
     
+    g_threads_count = numThreads < KS_MAX_THREADS_COUNT ? numThreads : KS_MAX_THREADS_COUNT;
+
+    for (mach_msg_type_number_t i = 0; i < numThreads; i++) {
```
Comment (bamx23) [inline Source/KSCrash/Recording/Tools/KSMachineContext.c]: `i < g_threads_count`?

Comment (JunyiXie) [reply Source/KSCrash/Recording/Tools/KSMachineContext.c]: Sorry, I fixed it.

**Comment by JunyiXie** [comment]
👍

**Comment by ZhiyuWong** [comment]
> Add a WeChat? @ZhiyuWong

OK

**Comment by ZhiyuWong** [comment]
> Use an array of size 1000 to store threads

How about use 
`static thread_act_array_t g_suspendedThreads = NULL;`
`static mach_msg_type_number_t g_suspendedThreadsCount = 0;`
instead of an arbitrary size of threads array.

**Comment by ZhiyuWong** [comment]
I think we can do it like this:

```
static thread_act_array_t g_suspendedThreads = NULL;
static mach_msg_type_number_t g_suspendedThreadsCount = 0;

void ksmc_suspendEnvironment()
{
#if KSCRASH_HAS_THREADS_API
    KSLOG_DEBUG("Suspending environment.");
    kern_return_t kr;
    const task_t thisTask = mach_task_self();
    const thread_t thisThread = (thread_t)ksthread_self();
    
    if((kr = task_threads(thisTask, &g_suspendedThreads, &g_suspendedThreadsCount)) != KERN_SUCCESS)
    {
        KSLOG_ERROR("task_threads: %s", mach_error_string(kr));
        return;
    }
    
    for(mach_msg_type_number_t i = 0; i < g_suspendedThreadsCount; i++)
    {
        thread_t thread = g_suspendedThreads[i];
        if(thread != thisThread && !isThreadInList(thread, g_reservedThreads, g_reservedThreadsCount))
        {
            if((kr = thread_suspend(thread)) != KERN_SUCCESS)
            {
                // Record the error and keep going.
                KSLOG_ERROR("thread_suspend (%08x): %s", thread, mach_error_string(kr));
            }
        }
    }

    KSLOG_DEBUG("Suspend complete.");
#endif
}

void ksmc_resumeEnvironment()
{
#if KSCRASH_HAS_THREADS_API
    KSLOG_DEBUG("Resuming environment.");
    kern_return_t kr;
    const task_t thisTask = mach_task_self();
    const thread_t thisThread = (thread_t)ksthread_self();
    
    if(g_suspendedThreads == NULL || g_suspendedThreadsCount == 0)
    {
        KSLOG_ERROR("we should call ksmc_suspendEnvironment() first");
        return;
    }
    
    for(mach_msg_type_number_t i = 0; i < g_suspendedThreadsCount; i++)
    {
        thread_t thread = g_suspendedThreads[i];
        if(thread != thisThread && !isThreadInList(thread, g_reservedThreads, g_reservedThreadsCount))
        {
            if((kr = thread_resume(thread)) != KERN_SUCCESS)
            {
                // Record the error and keep going.
                KSLOG_ERROR("thread_resume (%08x): %s", thread, mach_error_string(kr));
            }
        }
    }
    
    for(mach_msg_type_number_t i = 0; i < g_suspendedThreadsCount; i++)
    {
        mach_port_deallocate(thisTask, g_suspendedThreads[i]);
    }
    vm_deallocate(thisTask, (vm_address_t)g_suspendedThreads, sizeof(thread_t) * g_suspendedThreadsCount);
    g_suspendedThreads = NULL;
    g_suspendedThreadsCount = 0;

    KSLOG_DEBUG("Resume complete.");
#endif
}
```

**Comment by JunyiXie** [comment]
Great. Thanks
I want to release memory in suspend.
But that is  not  necessary, suspend and resume will be  called in pairs.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by JunyiXie** [comment]
> > Add a WeChat? @ZhiyuWong
> 
> OK

XiEJunYI_

**Pull request merged**